x/metadata/types/p8e.go

```diff
@@ -119,96 +119,159 @@ func convertP8eInputSpecs(old []*p8e.DefinitionSpec) (inputs []*InputSpecificati
+		}
+		facts[f.Name] = f
+	}
+
 	// Create the records.
+	p8EData.RecordReqs = make([]*RecordReq, 0, len(msg.Contract.Considerations))
 	for _, c := range msg.Contract.Considerations {
 		if c != nil && c.Result != nil && c.Result.Output != nil && c.Result.Result != p8e.ExecutionResultType_RESULT_TYPE_SKIP {
 			record := emptyRecord()
 			record.Name = c.ConsiderationName
 			record.SessionId = p8EData.Session.SessionId
-			record.Process.ProcessId = &Process_Hash{Hash: "NO-SPEC-HASH-AVAILABLE"}
+			record.Process.ProcessId = processID
 			record.Process.Name = c.Result.Output.Classname
 			record.Process.Method = record.Name
-			for _, f := range c.Inputs {
-				record.Inputs = append(record.Inputs, RecordInput{
+			record.Inputs = make([]RecordInput, len(c.Inputs))
+			for i, f := range c.Inputs {
+				if len(f.Hash) == 0 && len(f.Name) == 0 {
+					return p8EData, fmt.Errorf("consideration %s inputs[%d] name and hash cannot both be empty",
+						c.ConsiderationName, i)
+				}
+
+				ri := RecordInput{
 					Name:     f.Name,
-					Source:   &RecordInput_Hash{Hash: f.Hash},
 					TypeName: f.Classname,
 					Status:   RecordInputStatus_Unknown,
-				})
+				}
+
+				if len(f.Hash) > 0 {
+					ri.Status = RecordInputStatus_Proposed
+					ri.Source = &RecordInput_Hash{Hash: f.Hash}
+				} else {
+					ri.Status = RecordInputStatus_Record
+					if facts[f.Name] == nil {
+						return p8EData, fmt.Errorf("consideration %s inputs[%d] %s not found as contract input",
+							c.ConsiderationName, i, f.Name)
+					}
+					if facts[f.Name].DataLocation == nil || facts[f.Name].DataLocation.Ref == nil {
+						return p8EData, fmt.Errorf("contract input %s missing datalocation ref", f.Name)
+					}
+					if len(facts[f.Name].DataLocation.Ref.Hash) == 0 &&
+						(facts[f.Name].DataLocation.Ref.ScopeUuid == nil ||
+							len(facts[f.Name].DataLocation.Ref.ScopeUuid.Value) == 0) {
+						return p8EData, fmt.Errorf("contract input %s datalocation ref must have either a hash or scope uuid",
+							f.Name)
+					}
+
+					if len(facts[f.Name].DataLocation.Ref.Hash) > 0 {
+						ri.Source = &RecordInput_Hash{Hash: facts[f.Name].DataLocation.Ref.Hash}
+					} else {
+						scopeUUID, scopeUUIDErr := uuid.Parse(facts[f.Name].DataLocation.Ref.ScopeUuid.Value)
+						if scopeUUIDErr != nil {
+							return p8EData, fmt.Errorf("invalid UUID in contract input %s: %w", f.Name, scopeUUIDErr)
+						}
+						ri.Source = &RecordInput_RecordId{RecordId: RecordMetadataAddress(scopeUUID, f.Name)}
+					}
+				}
+
+				record.Inputs[i] = ri
 			}
 			record.Outputs = []RecordOutput{
 				{
 					Hash:   c.Result.Output.Hash,
 					Status: ResultStatus(c.Result.Result),
 				},
 			}
+			recReq := RecordReq{
+				Record:               record,
+				OriginalOutputHashes: []string{},
+			}
+			if c.Result.Output.Ancestor != nil && len(c.Result.Output.Ancestor.Hash) > 0 {
+				recReq.OriginalOutputHashes = append(recReq.OriginalOutputHashes, c.Result.Output.Ancestor.Hash)
```

Comment:
Again... changing from ancestor to original loses the superseded aspect of this naming convention...